### PR TITLE
bpo-30193: Allow to load buffer objects with json.loads()

### PR DESCRIPTION
**Pull request by fafhrd91**
it is not possible to load buffer object with json.loads()

### CODE REVIEW

File: Lib/json/__init__.py [hunk 1]
```diff
@@ -242,12 +242,13 @@ def dumps(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
 
 
 def detect_encoding(b):
-    bstartswith = b.startswith
-    if bstartswith((codecs.BOM_UTF32_BE, codecs.BOM_UTF32_LE)):
-        return 'utf-32'
-    if bstartswith((codecs.BOM_UTF16_BE, codecs.BOM_UTF16_LE)):
-        return 'utf-16'
-    if bstartswith(codecs.BOM_UTF8):
+    for prefix in (codecs.BOM_UTF32_BE, codecs.BOM_UTF32_LE):
```
Comment (DimitrisJim) [inline Lib/json/__init__.py:245]: Why was this change made?

Comment (fafhrd91) [reply Lib/json/__init__.py:245]: memoryview does not provide "startswith" function

Comment (DimitrisJim) [reply Lib/json/__init__.py:245]: D'oh, completely missed that, it's late here :) Anyway, support for bytes was added in https://bugs.python.org/issue17909 by Serhiy so it might make sense to ping him.

File: Lib/json/__init__.py [hunk 2]
```diff
@@ -242,12 +242,13 @@ def dumps(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
...
+        if b[0:len(prefix)] == prefix:
+            return 'utf-32'
+    for prefix in (codecs.BOM_UTF16_BE, codecs.BOM_UTF16_LE):
+        if b[0:len(prefix)] == prefix:
+            return 'utf-16'
+    if b[0:len(codecs.BOM_UTF8)] == codecs.BOM_UTF8:
```
Comment (vstinner) [inline Lib/json/__init__.py]: Replace [0:len(...)] with [:len(...)].

File: Lib/test/test_json/test_decode.py [hunk 3]
```diff
@@ -20,6 +21,15 @@ def test_empty_objects(self):
         self.assertEqual(self.loads('[]'), [])
         self.assertEqual(self.loads('""'), "")
 
+    def test_memoryview(self):
+        data = b'{"key": "val"}'
```
Comment (vstinner) [inline Lib/test/test_json/test_decode.py]: Please create the memoryview() on that line to make the test more explicit.

File: Misc/NEWS [hunk 4]
```diff
@@ -102,6 +102,8 @@ Core and Builtins
 
 - bpo-29546: Improve from-import error message with location
 
+- bpo-30193: Allow to load buffer objects with ``json.loads()``
```
Comment (vstinner) [inline Misc/NEWS]: The NEWS entry should be move at the top the Library section.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by fafhrd91** [comment]
@haypo updated according comments

File: Lib/json/__init__.py [hunk 5]
```diff
@@ -242,12 +242,13 @@ def dumps(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
...
 
 def detect_encoding(b):
-    bstartswith = b.startswith
-    if bstartswith((codecs.BOM_UTF32_BE, codecs.BOM_UTF32_LE)):
-        return 'utf-32'
-    if bstartswith((codecs.BOM_UTF16_BE, codecs.BOM_UTF16_LE)):
-        return 'utf-16'
-    if bstartswith(codecs.BOM_UTF8):
+    for prefix in (codecs.BOM_UTF32_BE, codecs.BOM_UTF32_LE):
+        if b[:len(prefix)] == prefix:
```
Comment (serhiy-storchaka) [inline Lib/json/__init__.py:246]: This doesn't work for e.g. `memoryview(array.array('u', '\ufeff[1,2,3]'))`.

Comment (fafhrd91) [reply Lib/json/__init__.py:246]: purpose of this change is to allow to use c-ext classes that supports buffer protocol.

I can remove memoryview from allowed type, and if encoding detection fails then fail with TypeError.

proper solution would be to write memoryview like object that supports only "b|B" format

Comment (fafhrd91) [reply Lib/json/__init__.py:246]: any conclusion on this ticket?

Comment (jakirkham) [reply Lib/json/__init__.py:246]: Is it reasonable to require unicode arrays for this patch? They are deprecated after all.

**Comment by jakirkham** [comment]
It looks like all comments were addressed or the OP was seeking further feedback from core on the requested changes. Would it be possible for someone from core to review this again?

File: Lib/json/__init__.py (hunk 5 above)
Comment (vstinner) [reply Lib/json/__init__.py:246]: I suggest to start simple and raise a TypeError if it's memory view with view.format != 'B'.

**Comment by flavianh** [comment]
@jakirkham  Revived in https://github.com/python/cpython/pull/14977

File: Misc/NEWS [hunk 6]
```diff
@@ -317,6 +317,8 @@ Extension Modules
 Library
 -------
 
+- bpo-30193: Allow to load buffer objects with ``json.loads()``
```
Comment (vstinner) [inline Misc/NEWS:320]: You should add a NEWS entry using the blurb tool, and revert this change.

Comment (jakirkham) [reply Misc/NEWS:320]: Looks like that is being done in @flavianh's follow-up PR ( https://github.com/python/cpython/pull/14977 ).

**Comment by csabella** [comment]
As this is from an unknown repository, I'm going to close this PR.  A[ replacement PR](https://github.com/python/cpython/pull/14977) has already been created.
